astropy/io/fits/column.py

```diff
@@ -1423,7 +1423,10 @@ def dtype(self):
                 else:
                     dt = np.dtype((dt.base, dim))
 
-            fields.append((name, dt))
+            if PY2:  # pragma: py2
+                fields.append((name.encode('utf-8'), dt))
```

Comment:
You tried with `u'Cœur'` which is unicode and non-ascii, but I was suggesting to try with `u'Cœur'.encode('utf8')` which is already encoded. It should issue a warning (from https://github.com/pllim/astropy/blob/fix-dtype-not-found/astropy/io/fits/column.py#L751), and fail in the line we are discussing.
Also you're right that `unicode_literals` is usually the thing that can cause unexpected unicode conversion (and that's why it's usage should be avoided if possible).
And I don't know well enough `io.fits` code to say what is the best fix (you could test `if PY2 and isinstance(name, unicde)`, or maybe there is a better place to do it), my point was just that encoding `str` strings is bad ;).